### PR TITLE
[mypy] Add mypy type annotation part 1

### PR DESCRIPTION
**Pull request by rkooo567**
This PR introduces mypy and fixes some issues from some of the directories. 

The goal

```
[done] mypy vllm/attention/*.py --follow-imports=skip --config-file pyproject.toml
[done] mypy vllm/core/*.py --follow-imports=skip --config-file pyproject.toml
[done] mypy vllm/distributed/*.py --follow-imports=skip --config-file pyproject.toml
[done] mypy vllm/engine/*.py --follow-imports=skip --config-file pyproject.toml
[done] mypy vllm/entrypoints/*.py --follow-imports=skip --config-file pyproject.toml
[done] mypy vllm/executor/*.py --follow-imports=skip --config-file pyproject.toml
[done] mypy vllm/usage/*.py --follow-imports=skip --config-file pyproject.toml
[done] mypy vllm/*.py --follow-imports=skip --config-file pyproject.toml
[done] mypy vllm/transformer_utils/*.py --follow-imports=skip --config-file pyproject.toml
```

Followup
```
mypy vllm/engine/*.py --follow-imports=skip --config-file pyproject.toml
mypy vllm/worker/*.py --follow-imports=skip --config-file pyproject.toml
mypy vllm/spec_decoding/*.py --follow-imports=skip --config-file pyproject.toml
mypy vllm/model_executor/*.py --follow-imports=skip --config-file pyproject.toml
mypy vllm/lora/*.py --follow-imports=skip --config-file pyproject.toml

```

For other files, I am planning to do it in a separate PR

Partially addressing https://github.com/vllm-project/vllm/issues/3680

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

File: pyproject.toml
Comment (michaelfeil) [inline pyproject.toml:46]: IIRC vllm's minimum version is still 3.8

File: vllm/core/scheduler.py
Comment (michaelfeil) [inline vllm/core/scheduler.py:599]: Is this type hint required?

**Review by michaelfeil — COMMENTED** [review]
Added some comments!

**Comment by DarkLight1337** [comment]
> @DarkLight1337 My mypy errors are all in slightly different files. Should be independent.

OP plans to commit further changes so it is better to notify in advance. Thanks for the speedy reply btw!

**Comment by michaelfeil** [comment]
I'll maybe check in with the above PR. Perhaps it makes sense to split up into multiple PRs to make it easier to review and less likley to have some conflicts that need to be resolved after approval.

File: pyproject.toml
Comment (WoosukKwon) [reply pyproject.toml:46]: +1. Please use python 3.8.

File: vllm/core/scheduler.py
Comment (rkooo567) [reply vllm/core/scheduler.py:341]: I found it has typing issue without this, I think because it thinks the return type is "deque"

File: vllm/core/scheduler.py
Comment (rkooo567) [reply vllm/core/scheduler.py:599]: I think we should use https://docs.python.org/3/library/typing.html#typing.Deque

I don't know if we should suppress this issue, but I think it has an issue because you don't know the element type, and mypy cannot automatically infer that?

File: vllm/sampling_params.py
Comment (rkooo567) [reply vllm/sampling_params.py:8]: let me fix this part

**Comment by rkooo567** [comment]
@michaelfeil so I guess we can merge your PR first and merge this PR as a follow up? since it introduces the mypy to the CI.

**Comment by rkooo567** [comment]
This is ready for the review.

File: vllm/core/scheduler.py
Comment (michaelfeil) [reply vllm/core/scheduler.py]: Ah, my bad. The len and type annotation perhaps hint a weird python implementation. Also interesting that __bool__ is not supported.

File: vllm/core/scheduler.py
Comment (michaelfeil) [reply vllm/core/scheduler.py:341]: But should be able to be simplified to 

`len(self.waiting) or len(self.running)  or len(self.swapped)` right?

File: vllm/sequence.py
Comment (michaelfeil) [inline vllm/sequence.py:367]: potentially not imported

File: vllm/config.py
Comment (simon-mo) [inline vllm/config.py:682]: i'm surprised you still need these while there's the if statement before. mypy should be catching these?

File: vllm/transformers_utils/tokenizer.py
Comment (simon-mo) [inline vllm/transformers_utils/tokenizer.py:31]: This might not be correct

**Review by simon-mo — APPROVED** [review]
lgtm overall but we should move run mypy to format.sh

**Comment by simon-mo** [comment]
Merged @michaelfeil's PR!

**Comment by DarkLight1337** [comment]
Just a heads-up, I have updated type annotations in `vllm.entrypoints.openai` via #4028.

File: vllm/transformers_utils/tokenizer.py
Comment (rkooo567) [reply vllm/transformers_utils/tokenizer.py:31]: just ignored type here

File: vllm/sequence.py
Comment (rkooo567) [reply vllm/sequence.py:367]: I think it is fixed with the other PR that's merged!

File: vllm/config.py
Comment (rkooo567) [reply vllm/config.py:682]: I think the if above is `speculative_model is None and num_speculative_tokens is None`, so it doesn't cover only one of them is None.

File: run_mypy.sh
Comment (rkooo567) [reply run_mypy.sh:1]: Moved back to format.sh. 

Btw, the reason why I made this file was to avoid duplication between format.sh and mypy.yaml. I just copied/pasted them there, but lmk if you have other good idea

File: vllm/core/scheduler.py
Comment (rkooo567) [reply vllm/core/scheduler.py:341]: actually it makes the return type confused with int. So I just kept the current one.

**Comment by rkooo567** [comment]
It is mergeable! I think the lora failure is unrelated

**Comment by rkooo567** [comment]
cc @simon-mo it is ready to merge!

**Pull request merged**
